src/Compilers/CSharp/Portable/Compilation/AttributeSemanticModel.cs

```diff
@@ -48,8 +48,29 @@ public static AttributeSemanticModel CreateSpeculative(SyntaxTreeSemanticModel p
             Debug.Assert(parentSemanticModel != null);
             Debug.Assert(rootBinder != null);
             Debug.Assert(rootBinder.IsSemanticModelBinder);
-
+            Debug.Assert(rootBinder.ContainingMember() is not SourcePropertyAccessorSymbol { Property.IsIndexer: false } || hasSpeculativeFieldKeywordBinderInChain(rootBinder));
```

Comment:
>Debug.Assert(rootBinder.ContainingMember() is not SourcePropertyAccessorSymbol { Property.IsIndexer: false } || hasSpeculativeFieldKeywordBinderInChain(rootBinder)); [](http://example.com/codeflow?start=12&length=165)

Consider moving this assert to the only ```AttributeSemanticModel``` constructor. It can detect whether we are speculating by checking ```parentSemanticModelOpt is not null```.